roachprod: distinguish infra and crdb errors

jlinder opened the pull request:
Before: there are many reasons why roachprod might fail for a given command
and it didn't differentiate between the different types of errors.

Why change?

- The lack of detail for why an error happened made debugging problems harder
  than needed.

- We couldn't use automated methods to aggregate error reporting or direct
  errors to the right team when it's used by roachtest.

- We want to gather metrics on specific types of errors and it's wasn't
  possible without more granular error reporting from roachprod.

Now:

- roachprod exits with different codes depending on type of error:

    0: all OK
    1: an unclassified roachprod error
    10: a problem with an SSH connection to a server in the cluster
    20: a problem running a non-cockroach command on a remote server in the
       cluster
    30: a problem running a cockroach command on a remote server in the cluster

- for each error type, an easily-searchable string is emitted by roachprod at
  or near the end of output. The string will also be emitted for each error
  that happens during an ssh connection to a remote cluster node. The strings
  for each error code are:

    1: UNCLASSIFIED_PROBLEM
    10: SSH_PROBLEM
    20: COMMAND_PROBLEM
    30: DEAD_ROACH_PROBLEM

- Refined error handling for remote ssh connections only for commands that use
  SyncedCluster.Run().

Release justification: This dosn't affect CRDB itself, only internal tools.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/46201)
<!-- Reviewable:end -->

jlinder commented on pkg/cmd/roachprod/install/cluster_synced.go:
```diff
@@ -491,14 +521,26 @@ func (c *SyncedCluster) Run(stdout, stderr io.Writer, nodes []int, title, cmd st
 		}
 	}
 
+	// TODO(jamesl): Choose the best error to return. SSH? cockroach? CMD? roachprod?
```
By providing a deterministic hierarchy of result error types, it could make debugging easier. This is particularly relevant to roachtest. Is this desirable? Is there a best priority order? My guess would be: cockroach, ssh, roachprod, other cmd.

ajwerner reviewed (commented):
I'll take another pass in a few

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-:-M2dRHJKDJls0RAS1Z4T:b-gkkqss)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @bobvawter, @jlinder, and @tbg)

___
*[pkg/cmd/roachprod/main.go, line 32 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2dQDu4Df6w9Wwjps0q:-M2dQDu51gYD7ZpjHwPr:b-xk3qiz) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd0513b4dab788553af31a2984b4f508de47ca39/pkg/cmd/roachprod/main.go#L32)):*
> ```Go
> 	cld "github.com/cockroachdb/cockroach/pkg/cmd/roachprod/cloud"
> 	"github.com/cockroachdb/cockroach/pkg/cmd/roachprod/config"
> 	rpErrors "github.com/cockroachdb/cockroach/pkg/cmd/roachprod/errors"
> ```

nit: don't camelCase package names. `rperrors` is fine. 
___
*[pkg/cmd/roachprod/main.go, line 270 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2dPcdk24_Aj1XJHi9Y:-M2dPcdlCiBHUDNlWm-c:boio5m4) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd0513b4dab788553af31a2984b4f508de47ca39/pkg/cmd/roachprod/main.go#L270)):*
> ```Go
> 		err := f(cmd, args)
> 		if err != nil {
> 			roachprodError, ok := err.(rpErrors.BaseError)
> ```

Have you seen https://github.com/cockroachdb/errors? 

I think this library will be helpful for extracting errors.


Probably write a function to extract the error in the `roachprod/errors` package:

```
package errors

import crdberrors "github.com/cockroachdb/errors"

func AsError(err error) (Error, bool) {
    if rpErr, ok := crdberrors.If(err, func(err error) (interface{}, bool) {
        if rpErr, ok := err.(Error); ok {
            return rpErr, true
        }
        return nil, false
    }); ok {
        return rpErr.(Error), true
    }
    return nil, false
}
```
___
*[pkg/cmd/roachprod/errors/errors.go, line 20 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2dQIMc4dmIwqMNVXLy:-M2dQIMc4dmIwqMNVXLz:b-x0zh6q) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd0513b4dab788553af31a2984b4f508de47ca39/pkg/cmd/roachprod/errors/errors.go#L20)):*
> ```Go
> // BaseError: Errors implementing this interface are used by the wrap() function in
> // the main roachprod package to output correctly classified log messages and exit codes.
> type BaseError interface {
> ```

nit: just call this `Error`


<!-- Sent from Reviewable.io -->

bobvawter reviewed (changes requested):
> Reviewed 6 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46201)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jlinder and @tbg)

___
*[pkg/cmd/roachprod/README.md, line 185 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2e08gE4kXPHSJuVEKb:-M2e08gE4kXPHSJuVEKc:bc46fk8) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd0513b4dab788553af31a2984b4f508de47ca39/pkg/cmd/roachprod/README.md#L185)):*
> ```Markdown
> - 2: a problem running a non-cockroach command on a remote cluster server or on a local node
> - 3: a problem running a cockroach command on a remote cluster server or a local node
> - 5: a problem with an SSH connection to a server in the cluster
> ```

What happened to 4?
___
*[pkg/cmd/roachprod/errors/errors.go, line 22 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2dzx9UBiNei45PcRP0:-M2dzx9UBiNei45PcRP1:b-wqyuze) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd0513b4dab788553af31a2984b4f508de47ca39/pkg/cmd/roachprod/errors/errors.go#L22)):*
> ```Go
> type BaseError interface {
> 	Error() string
> 	LogString() string
> ```

Document how `LogString()` differs from `Error()`?

Will this also tell me the specific command line that failed?
___
*[pkg/cmd/roachprod/errors/errors.go, line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2dzuh47k9WpLJa4CAX:-M2dzuh47k9WpLJa4CAY:blcm2jc) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd0513b4dab788553af31a2984b4f508de47ca39/pkg/cmd/roachprod/errors/errors.go#L23)):*
> ```Go
> 	Error() string
> 	LogString() string
> 	ReturnCode() int
> ```

`ExitCode()` ?
___
*[pkg/cmd/roachprod/errors/errors.go, line 118 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2e-Rcl4GPvnaheCPk4:-M2e-Rcl4GPvnaheCPk5:b-s526v9) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd0513b4dab788553af31a2984b4f508de47ca39/pkg/cmd/roachprod/errors/errors.go#L118)):*
> ```Go
> // Classify an error received while executing a non-cockroach command remotely
> // over an ssh connection to the right BaseError type.
> func ClassifyCmdError(err error) error {
> ```

Should the return type for these `Classify` functions be `BaseError` (or just `Error`)?
___
*[pkg/cmd/roachprod/install/cluster_synced.go, line 89 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2dsmKH5a22zQhw0IfA:-M2dsmKH5a22zQhw0IfB:b7ewpg5) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd0513b4dab788553af31a2984b4f508de47ca39/pkg/cmd/roachprod/install/cluster_synced.go#L89)):*
> ```Go
> 	CockroachCmd = 1
> 	OtherCmd = 2
> )
> ```

This would usually be written as

```
const (
  CockroachCmd CmdKind = iota // or iota+1 if you want to start at 1
  OtherCmd // This will implicitly be of the same kind and incremented
)
```

The reason to provide an explicit type is to prevent golang's usual behavior of allowing arbitrary reinterpretation of constant, untyped ints.

X-Ref: https://golang.org/ref/spec#Constant_declarations
___
*[pkg/cmd/roachprod/install/cluster_synced.go, line 502 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2e-ohi2gqpMVF6Z-mL:-M2e-ohi2gqpMVF6Z-mM:bte8uwh) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd0513b4dab788553af31a2984b4f508de47ca39/pkg/cmd/roachprod/install/cluster_synced.go#L502)):*
> ```Go
> 					errors[i] = rpErrors.ClassifyCockroachError(errors[i])
> 				} else {
> 					errors[i] = rpErrors.ClassifyCmdError(errors[i])
> ```

Would it make sense to plumb `cmdKind` through to a unified `ClassifyError()` function?
___
*[pkg/cmd/roachprod/install/cluster_synced.go, line 537 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2dPJmHMBubImK4r9Xh-r1-537:-M2e00CF6X23cqL16CgI:bz9sdox) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd0513b4dab788553af31a2984b4f508de47ca39/pkg/cmd/roachprod/install/cluster_synced.go#L537)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

By providing a deterministic hierarchy of result error types, it could make debugging easier. This is particularly relevant to roachtest. Is this desirable? Is there a best priority order? My guess would be: cockroach, ssh, roachprod, other cmd.
</blockquote></details>

That seems like a reasonable approach.
___
*[pkg/cmd/roachprod/install/cluster_synced.go, line 543 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2e0Kkd33zWqxLr6HNz:-M2e0Kkd33zWqxLr6HO-:b1t2i16) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd0513b4dab788553af31a2984b4f508de47ca39/pkg/cmd/roachprod/install/cluster_synced.go#L543)):*
> ```Go
> 		}
> 	}
> ```

Misc whitespace change.


<!-- Sent from Reviewable.io -->

jlinder reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46201)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @bobvawter, and @tbg)

___
*[pkg/cmd/roachprod/main.go, line 270 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2dPcdk24_Aj1XJHi9Y:-M2iDiBh2w01obAzKwZG:bti0rqb) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd0513b4dab788553af31a2984b4f508de47ca39/pkg/cmd/roachprod/main.go#L270)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Have you seen https://github.com/cockroachdb/errors? 

I think this library will be helpful for extracting errors.


Probably write a function to extract the error in the `roachprod/errors` package:

```
package errors

import crdberrors "github.com/cockroachdb/errors"

func AsError(err error) (Error, bool) {
    if rpErr, ok := crdberrors.If(err, func(err error) (interface{}, bool) {
        if rpErr, ok := err.(Error); ok {
            return rpErr, true
        }
        return nil, false
    }); ok {
        return rpErr.(Error), true
    }
    return nil, false
}
```
</blockquote></details>

I have heard of `cockroachdb/errors` but had not yet investigated it. Thanks for pointing it out. Looking at it now.
___
*[pkg/cmd/roachprod/README.md, line 185 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2e08gE4kXPHSJuVEKb:-M2e46jgD41-Gy_c82HM:buv38sp) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd0513b4dab788553af31a2984b4f508de47ca39/pkg/cmd/roachprod/README.md#L185)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

What happened to 4?
</blockquote></details>

The initial reason is that, because the SSH connection error is a pretty specific error, putting it with a higher number that would allow other, very specific or related errors to congregate around would be nice. And I thought there would be only 3 codes, so 1, 2 and 5 would work. Then the second code split in two and I used 1, 2, 7, and 8 for a while, but then switched to this lineup which leaves 5 still with some space around it (and is the start of the fibonacci sequence). 

What are your thoughts on how to assign return codes (particularly where changing them later could be challenging since other programs rely on them)?
___
*[pkg/cmd/roachprod/errors/errors.go, line 22 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2dzx9UBiNei45PcRP0:-M2iEXAh5PWE_P-I0jiL:b-vs4pco) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd0513b4dab788553af31a2984b4f508de47ca39/pkg/cmd/roachprod/errors/errors.go#L22)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

Document how `LogString()` differs from `Error()`?

Will this also tell me the specific command line that failed?
</blockquote></details>

Thank you for asking. `LogString()` only returns the well-known searchable string. `Error()` incorporates the value from `LogString()`. Getting the specific command that failed is an interesting idea. It doesn't seem easy to do (`set -x` is possible but gives a lot more output than wanted normally).

I'm evaluating this breakdown given @ajwerner's pointer to `cockroachdb/errors`.
___
*[pkg/cmd/roachprod/errors/errors.go, line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2dzuh47k9WpLJa4CAX:-M2iO-4W35z_CyNejsdI:btx61kn) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd0513b4dab788553af31a2984b4f508de47ca39/pkg/cmd/roachprod/errors/errors.go#L23)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

`ExitCode()` ?
</blockquote></details>

Added a description comment.
___
*[pkg/cmd/roachprod/errors/errors.go, line 118 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2e-Rcl4GPvnaheCPk4:-M2e6MjM73rlZ3jIiYLa:b-v54mth) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd0513b4dab788553af31a2984b4f508de47ca39/pkg/cmd/roachprod/errors/errors.go#L118)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

Should the return type for these `Classify` functions be `BaseError` (or just `Error`)?
</blockquote></details>

Yes. Changed to `Error`.
___
*[pkg/cmd/roachprod/install/cluster_synced.go, line 89 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2dsmKH5a22zQhw0IfA:-M2e7FNy-Mu1hINmFz_U:b-x3zn47) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd0513b4dab788553af31a2984b4f508de47ca39/pkg/cmd/roachprod/install/cluster_synced.go#L89)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

This would usually be written as

```
const (
  CockroachCmd CmdKind = iota // or iota+1 if you want to start at 1
  OtherCmd // This will implicitly be of the same kind and incremented
)
```

The reason to provide an explicit type is to prevent golang's usual behavior of allowing arbitrary reinterpretation of constant, untyped ints.

X-Ref: https://golang.org/ref/spec#Constant_declarations
</blockquote></details>

:thumbsup: Done.
___
*[pkg/cmd/roachprod/install/cluster_synced.go, line 502 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2e-ohi2gqpMVF6Z-mL:-M2e7c-S-uvQpTFLfk1E:b-8sp4zu) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd0513b4dab788553af31a2984b4f508de47ca39/pkg/cmd/roachprod/install/cluster_synced.go#L502)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

Would it make sense to plumb `cmdKind` through to a unified `ClassifyError()` function?
</blockquote></details>

Possibly. I had chosen not to do that in `pkg/cmd/roachprod/errors` because the concept of `CmdKind` isn't a global-error-related concept and mapping a `CmdKind` to an `ErrorKind` would require lining up the values of the kinds or some other mapping mechanism (which I imagine would be ugly and brittleish).

Splitting out the function locally would work. And then, if the logic applies to the other places in the code making ssh calls, I/we could figure out how make it a little more generic for the new location at that time.

...

I added a method on `CmdKind`. It did make the code nicer.
___
*[pkg/cmd/roachprod/install/cluster_synced.go, line 537 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2dPJmHMBubImK4r9Xh-r1-537:-M2iPqtuEQlRD5YjGdvG:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd0513b4dab788553af31a2984b4f508de47ca39/pkg/cmd/roachprod/install/cluster_synced.go#L537)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

That seems like a reasonable approach.
</blockquote></details>

:thumbsup: 
___
*[pkg/cmd/roachprod/install/cluster_synced.go, line 543 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2e0Kkd33zWqxLr6HNz:-M2eLfNFBov9VN_wFClG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd0513b4dab788553af31a2984b4f508de47ca39/pkg/cmd/roachprod/install/cluster_synced.go#L543)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

Misc whitespace change.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

jlinder reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46201)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @bobvawter, and @tbg)

___
*[pkg/cmd/roachprod/errors/errors.go, line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2dzuh47k9WpLJa4CAX:-M2iRjJZD03s7Ceg-MI7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd0513b4dab788553af31a2984b4f508de47ca39/pkg/cmd/roachprod/errors/errors.go#L23)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

Added a description comment.
</blockquote></details>

Done.
___
*[pkg/cmd/roachprod/errors/errors.go, line 118 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2e-Rcl4GPvnaheCPk4:-M2iRhPyBI9lDiw6qXGE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd0513b4dab788553af31a2984b4f508de47ca39/pkg/cmd/roachprod/errors/errors.go#L118)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

Yes. Changed to `Error`.
</blockquote></details>

Done.
___
*[pkg/cmd/roachprod/install/cluster_synced.go, line 89 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2dsmKH5a22zQhw0IfA:-M2iRfK_BtI5jeaHwjCj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd0513b4dab788553af31a2984b4f508de47ca39/pkg/cmd/roachprod/install/cluster_synced.go#L89)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

:thumbsup: Done.
</blockquote></details>

Done.
___
*[pkg/cmd/roachprod/install/cluster_synced.go, line 502 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2e-ohi2gqpMVF6Z-mL:-M2iRckhBGH0YmkqR35d:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd0513b4dab788553af31a2984b4f508de47ca39/pkg/cmd/roachprod/install/cluster_synced.go#L502)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

Possibly. I had chosen not to do that in `pkg/cmd/roachprod/errors` because the concept of `CmdKind` isn't a global-error-related concept and mapping a `CmdKind` to an `ErrorKind` would require lining up the values of the kinds or some other mapping mechanism (which I imagine would be ugly and brittleish).

Splitting out the function locally would work. And then, if the logic applies to the other places in the code making ssh calls, I/we could figure out how make it a little more generic for the new location at that time.

...

I added a method on `CmdKind`. It did make the code nicer.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

jlinder reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46201)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @bobvawter, and @tbg)

___
*[pkg/cmd/roachprod/main.go, line 270 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2dPcdk24_Aj1XJHi9Y:-M2jf_hU8X8HvtExWf3e:b-w71ykw) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd0513b4dab788553af31a2984b4f508de47ca39/pkg/cmd/roachprod/main.go#L270)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

I have heard of `cockroachdb/errors` but had not yet investigated it. Thanks for pointing it out. Looking at it now.
</blockquote></details>

Changed. Now it'll handle the case where the original raised error is wrapped further down the stack and outputs the full error message including any additional wrapped messages. 
___
*[pkg/cmd/roachprod/errors/errors.go, line 22 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2dzx9UBiNei45PcRP0:-M2jlkNlD5J5KE-AGEPh:b5zakdh) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd0513b4dab788553af31a2984b4f508de47ca39/pkg/cmd/roachprod/errors/errors.go#L22)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

Thank you for asking. `LogString()` only returns the well-known searchable string. `Error()` incorporates the value from `LogString()`. Getting the specific command that failed is an interesting idea. It doesn't seem easy to do (`set -x` is possible but gives a lot more output than wanted normally).

I'm evaluating this breakdown given @ajwerner's pointer to `cockroachdb/errors`.
</blockquote></details>

Now that I have a better understanding of what idiomatic error handling is like after digesting `cockroachdb/errors`, I've removed `LogString()` and left just `Error()`. I haven't added a mechanism to get more details, like the command that failed. Your question indicated it may be useful, though I haven't seen how this current output would appear in real logs and what other context is already there. Would it be useful?
___
*[pkg/cmd/roachprod/errors/errors.go, line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2jic9z0X6Z0noRz8sj:-M2jic9z0X6Z0noRz8sk:bokzb7v) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a3090cfb1b940e78c7e2911d86b40e9a767de12/pkg/cmd/roachprod/errors/errors.go#L35)):*
> ```Go
> //
> // For errors coming from a cockroach command, use CockroachError.
> type CmdError struct {
> ```

@bobvawter @ajwerner What is your take on renaming `CmdError` to just `Cmd` because it's in the `roachprod.errors` package? (Asking because I saw mention that naming errors within an errors package without the appended `Error` is preferred but it wasn't brought up.)


<!-- Sent from Reviewable.io -->

bobvawter reviewed (changes requested):
It looks like you've had some random generated files creep into the PR.

> Reviewed 2 of 5 files at r2, 3 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-:-M2n9wB294s89jbyQQmo:b-80duaw)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @jlinder, and @tbg)

___
*[pkg/cmd/roachprod/README.md, line 185 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2e08gE4kXPHSJuVEKb:-M2n8c9i1fTTLyo9v72m:bkdbh2m) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd0513b4dab788553af31a2984b4f508de47ca39/pkg/cmd/roachprod/README.md#L185)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

The initial reason is that, because the SSH connection error is a pretty specific error, putting it with a higher number that would allow other, very specific or related errors to congregate around would be nice. And I thought there would be only 3 codes, so 1, 2 and 5 would work. Then the second code split in two and I used 1, 2, 7, and 8 for a while, but then switched to this lineup which leaves 5 still with some space around it (and is the start of the fibonacci sequence). 

What are your thoughts on how to assign return codes (particularly where changing them later could be challenging since other programs rely on them)?
</blockquote></details>

I don't see much value in over-designing the exit codes without some specific use-case in mind.  If you want to leave room for further "minor codes", you could space the values out by 10's.  I just saw a gap in the numbers and was curious.
___
*[pkg/cmd/roachprod/errors/errors.go, line 22 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2dzx9UBiNei45PcRP0:-M2n7T2k6XAMD14fpnJP:b-q4dbm6) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd0513b4dab788553af31a2984b4f508de47ca39/pkg/cmd/roachprod/errors/errors.go#L22)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

Now that I have a better understanding of what idiomatic error handling is like after digesting `cockroachdb/errors`, I've removed `LogString()` and left just `Error()`. I haven't added a mechanism to get more details, like the command that failed. Your question indicated it may be useful, though I haven't seen how this current output would appear in real logs and what other context is already there. Would it be useful?
</blockquote></details>

If it's possible to get the actual command without turning this into a science project, it might help whomever is trying to debug an issue after-the-fact by hopefully bubbling up the "context" of the error message.
___
*[pkg/cmd/roachprod/errors/errors.go, line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2jic9z0X6Z0noRz8sj:-M2n7rrs-5VjNHKGO6T7:bqce6l7) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a3090cfb1b940e78c7e2911d86b40e9a767de12/pkg/cmd/roachprod/errors/errors.go#L35)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

@bobvawter @ajwerner What is your take on renaming `CmdError` to just `Cmd` because it's in the `roachprod.errors` package? (Asking because I saw mention that naming errors within an errors package without the appended `Error` is preferred but it wasn't brought up.)
</blockquote></details>

I don't have a strong opinion.  Perhaps I'm just used to go's verbosity.


<!-- Sent from Reviewable.io -->

jlinder committed:
roachprod: distinguish infra and crdb errors

Before: there are many reasons why roachprod might fail for a given command
and it didn't differentiate between the different types of errors.

Why change?

- The lack of detail for why an error happened made debugging problems harder
  than needed.

- We couldn't use automated methods to aggregate error reporting or direct
  errors to the right team when it's used by roachtest.

- We want to gather metrics on specific types of errors and it's wasn't
  possible without more granular error reporting from roachprod.

Now:

- roachprod exits with different codes depending on type of error:

    0: all OK
    1: an unclassified roachprod error
    10: a problem with an SSH connection to a server in the cluster
    20: a problem running a non-cockroach command on a remote server in the
       cluster
    30: a problem running a cockroach command on a remote server in the cluster

- for each error type, an easily-searchable string is emitted by roachprod at
  or near the end of output. The string will also be emitted for each error
  that happens during an ssh connection to a remote cluster node. The strings
  for each error code are:

    1:  UNCLASSIFIED_PROBLEM
    10: SSH_PROBLEM
    20: COMMAND_PROBLEM
    30: DEAD_ROACH_PROBLEM

- Refined error handling for remote ssh connections only for commands that use
  SyncedCluster.Run().

Release justification: This dosn't affect CRDB itself, only internal tools.

Release note: None

ajwerner reviewed (changes requested):
> Reviewed 1 of 6 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46201)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @bobvawter, @jlinder, and @tbg)

___
*[pkg/cmd/roachprod/errors/errors.go, line 21 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2iVrej1DJCFLg7UeMi:-M2iVrek0eGJE8Bp8NI4:bim585u) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a3090cfb1b940e78c7e2911d86b40e9a767de12/pkg/cmd/roachprod/errors/errors.go#L21)):*
> ```Go
> // the main roachprod package to output correctly classified log messages and exit codes.
> type Error interface {
> 	Error() string
> ```

nit: embed `error` instead of this. 
___
*[pkg/cmd/roachprod/errors/errors.go, line 28 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2iXBc308ucazy-4jML:-M2iXBc308ucazy-4jMM:b-5prqhn) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a3090cfb1b940e78c7e2911d86b40e9a767de12/pkg/cmd/roachprod/errors/errors.go#L28)):*
> ```Go
> 
> 	// Get the wrapped error.
> 	Unwrap() error
> ```

The errors library will deal with this transparently. I don't think this needs to be on the interface. Think of this interface as a description of a normal error. Error wrapping has become standard in go. 
___
*[pkg/cmd/roachprod/errors/errors.go, line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2jic9z0X6Z0noRz8sj:-M2o4pB02OB0FokPZjOT:bfhu8rq) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a3090cfb1b940e78c7e2911d86b40e9a767de12/pkg/cmd/roachprod/errors/errors.go#L35)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

I don't have a strong opinion.  Perhaps I'm just used to go's verbosity.
</blockquote></details>

nit: don't export the wrapped error, just have the caller use `Unwrap`.
___
*[pkg/cmd/roachprod/errors/errors.go, line 112 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2o5m2v53H_VctoTikp:-M2o5m2v53H_VctoTikq:b-i0xltt) ([raw file](https://github.com/cockroachdb/cockroach/blob/19a53c71aed1894b05f5d974bd38aa5832bf77aa/pkg/cmd/roachprod/errors/errors.go#L112)):*
> ```Go
> 	}
> 
> 	if exitErr, ok := err.(*exec.ExitError); ok {
> ```

use the errors library to see if there's an exec error inside of this error. 
___
*[pkg/cmd/roachprod/errors/errors.go, line 130 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2o5pmpAcs2omjrpeqh:-M2o5pmpAcs2omjrpeqi:b-i0xltt) ([raw file](https://github.com/cockroachdb/cockroach/blob/19a53c71aed1894b05f5d974bd38aa5832bf77aa/pkg/cmd/roachprod/errors/errors.go#L130)):*
> ```Go
> 	}
> 
> 	if exitErr, ok := err.(*exec.ExitError); ok {
> ```

use the errors library to see if there's an exec error inside of this error. 
___
*[pkg/cmd/roachprod/errors/errors.go, line 167 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2o5s5fEX-1ezStR_F1:-M2o5s5fEX-1ezStR_F2:bdyd15w) ([raw file](https://github.com/cockroachdb/cockroach/blob/19a53c71aed1894b05f5d974bd38aa5832bf77aa/pkg/cmd/roachprod/errors/errors.go#L167)):*
> ```Go
> 		}
> 
> 		rpErr, ok := err.(Error)
> ```

use `AsError` here
___
*[pkg/cmd/roachprod/errors/errors.go, line 186 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2o61U23NL_hEoGl-1F:-M2o61U23NL_hEoGl-1G:b-8se8jy) ([raw file](https://github.com/cockroachdb/cockroach/blob/19a53c71aed1894b05f5d974bd38aa5832bf77aa/pkg/cmd/roachprod/errors/errors.go#L186)):*
> ```Go
> 		return result
> 	}
> 	return backupResult
> ```

This code might be easier to read if you instead had another for loop here for the case where you didn't find an error from this package but still want to check if there's a non-nil error in `errors`.
___
*[pkg/cmd/roachprod/install/cluster_synced.go, line 508 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2o5C94DWeU7UgmG1cE:-M2o5C94DWeU7UgmG1cF:b-391tg1) ([raw file](https://github.com/cockroachdb/cockroach/blob/19a53c71aed1894b05f5d974bd38aa5832bf77aa/pkg/cmd/roachprod/install/cluster_synced.go#L508)):*
> ```Go
> 			if errors[i] != nil {
> 				err := cmdKind.classifyError(errors[i])
> 				fmt.Fprintln(stdout, err.Error())
> ```

my sense is that you'll want to contextualize this line of output, perhaps with the node on which the command was being run and also with the command which failed.

I'm not sure we want to add this logging at all. Furthermore, stderr feels more appropriate if we do. 
___
*[pkg/cmd/roachprod/install/cluster_synced.go, line 529 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2o5Wl4218zP8L3Y4Au:-M2o5Wl4218zP8L3Y4Av:b-ia41wb) ([raw file](https://github.com/cockroachdb/cockroach/blob/19a53c71aed1894b05f5d974bd38aa5832bf77aa/pkg/cmd/roachprod/install/cluster_synced.go#L529)):*
> ```Go
> 			if errors[i] != nil {
> 				err := cmdKind.classifyError(errors[i])
> 				fmt.Fprintln(stdout, err.Error())
> ```

same comment as above
___
*[pkg/storage/enginepb/mvcc3.pb.go, line 0 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2o5_MM6y18xVsPqRTV:-M2o5_MM6y18xVsPqRTW:b-jwfafa) ([raw file](https://github.com/cockroachdb/cockroach/blob/19a53c71aed1894b05f5d974bd38aa5832bf77aa/pkg/storage/enginepb/mvcc3.pb.go#L0)):*
This seems erroneous. 


<!-- Sent from Reviewable.io -->

jlinder reviewed (commented):
Ah! Thanks for pointing out. They are removed.

I didn't expect files updated / generated during `make bin/roachprod` to be checked into the repo. So when I ammended changes to the commit, they must have slipped in. And then I didn't register they were not supposed to be there when looking at the PR itself.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-:-M2nHCzV47ZKOOZG48Qn:b-mbbhwk)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @bobvawter, and @tbg)

___
*[pkg/cmd/roachprod/README.md, line 185 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2e08gE4kXPHSJuVEKb:-M2oodYmBHy3B11yURYK:b-w27uvz) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd0513b4dab788553af31a2984b4f508de47ca39/pkg/cmd/roachprod/README.md#L185)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

I don't see much value in over-designing the exit codes without some specific use-case in mind.  If you want to leave room for further "minor codes", you could space the values out by 10's.  I just saw a gap in the numbers and was curious.
</blockquote></details>

I spaced them out by 10. And used the values in `SelectPriorityError` below.
___
*[pkg/cmd/roachprod/errors/errors.go, line 22 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2dzx9UBiNei45PcRP0:-M2op3oj87X5LWRDtI9T:bew6jqm) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd0513b4dab788553af31a2984b4f508de47ca39/pkg/cmd/roachprod/errors/errors.go#L22)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

If it's possible to get the actual command without turning this into a science project, it might help whomever is trying to debug an issue after-the-fact by hopefully bubbling up the "context" of the error message.
</blockquote></details>

It was easy to get the full command that was run. But not the individual place where it failed.
___
*[pkg/cmd/roachprod/errors/errors.go, line 21 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2iVrej1DJCFLg7UeMi:-M2oqe4T9N68KQt2_Gnr:bhai4g2) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a3090cfb1b940e78c7e2911d86b40e9a767de12/pkg/cmd/roachprod/errors/errors.go#L21)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit: embed `error` instead of this. 
</blockquote></details>

I think I did what you asked. Did I interpret correctly?
___
*[pkg/cmd/roachprod/errors/errors.go, line 28 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2iXBc308ucazy-4jML:-M2oqJrzDpH-Y1ubiKZS:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a3090cfb1b940e78c7e2911d86b40e9a767de12/pkg/cmd/roachprod/errors/errors.go#L28)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

The errors library will deal with this transparently. I don't think this needs to be on the interface. Think of this interface as a description of a normal error. Error wrapping has become standard in go. 
</blockquote></details>

Removed.
___
*[pkg/cmd/roachprod/errors/errors.go, line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2jic9z0X6Z0noRz8sj:-M2nrCtL9sbsShK55jV4:b2laz1l) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a3090cfb1b940e78c7e2911d86b40e9a767de12/pkg/cmd/roachprod/errors/errors.go#L35)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit: don't export the wrapped error, just have the caller use `Unwrap`.
</blockquote></details>

I took off the trailing `Error` from all four of the errors. Following the logic of: the package is called `errors` and the types in the package are all types of errors + they are only ever used with a reference to the errors package (or within the package itself) and have that context thus there is no need for duplicating the word 'error' in the name.
___
*[pkg/cmd/roachprod/errors/errors.go, line 112 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2o5m2v53H_VctoTikp:-M2opMpjDdN_bd5Lio9W:b5np20g) ([raw file](https://github.com/cockroachdb/cockroach/blob/19a53c71aed1894b05f5d974bd38aa5832bf77aa/pkg/cmd/roachprod/errors/errors.go#L112)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

use the errors library to see if there's an exec error inside of this error. 
</blockquote></details>

Oh, right. Done.
___
*[pkg/cmd/roachprod/errors/errors.go, line 130 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2o5pmpAcs2omjrpeqh:-M2opVPd5Om0UlVtUOIL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/19a53c71aed1894b05f5d974bd38aa5832bf77aa/pkg/cmd/roachprod/errors/errors.go#L130)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

use the errors library to see if there's an exec error inside of this error. 
</blockquote></details>

Done.
___
*[pkg/cmd/roachprod/errors/errors.go, line 167 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2o5s5fEX-1ezStR_F1:-M2opoh221zlRb65yggr:b5np20g) ([raw file](https://github.com/cockroachdb/cockroach/blob/19a53c71aed1894b05f5d974bd38aa5832bf77aa/pkg/cmd/roachprod/errors/errors.go#L167)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

use `AsError` here
</blockquote></details>

Oh, right. Done.
___
*[pkg/cmd/roachprod/errors/errors.go, line 186 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2o61U23NL_hEoGl-1F:-M2oiyQa9ZxgN44M8J68:b38r5ye) ([raw file](https://github.com/cockroachdb/cockroach/blob/19a53c71aed1894b05f5d974bd38aa5832bf77aa/pkg/cmd/roachprod/errors/errors.go#L186)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

This code might be easier to read if you instead had another for loop here for the case where you didn't find an error from this package but still want to check if there's a non-nil error in `errors`.
</blockquote></details>

Made the change. It is easier to read now.
___
*[pkg/cmd/roachprod/install/cluster_synced.go, line 537 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2dPJmHMBubImK4r9Xh-r1-537:-M2nr4Sy6jF-5bRVMQ3W:bgw8bs) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd0513b4dab788553af31a2984b4f508de47ca39/pkg/cmd/roachprod/install/cluster_synced.go#L537)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

:thumbsup: 
</blockquote></details>

Completed.
___
*[pkg/cmd/roachprod/install/cluster_synced.go, line 508 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2o5C94DWeU7UgmG1cE:-M2oAfZWCBmC24nUQtyd:b-a9bu6q) ([raw file](https://github.com/cockroachdb/cockroach/blob/19a53c71aed1894b05f5d974bd38aa5832bf77aa/pkg/cmd/roachprod/install/cluster_synced.go#L508)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

my sense is that you'll want to contextualize this line of output, perhaps with the node on which the command was being run and also with the command which failed.

I'm not sure we want to add this logging at all. Furthermore, stderr feels more appropriate if we do. 


</blockquote></details>

Great thoughts.

Using `stderr` here is likely the right move because this is error information and not regular logging. Does roachtest capture stderr output in the correct place in the logs from roachprod runs (including placing it contextually where the message happens)? If not, we'll want to update roachtest to capture them correctly. (Because adding these messages is specifically done to be captured in the roachtest logs.) 

I added a `WithDetail` containing the node number and the command that was run. Getting that info all output somewhere seems like it would be helpful for debugging (particularly for roachtest). Since this line only gets called when the command runs on one node, outputting the error info logically would be done in the Command wrapper in roachprod main. However, for the case below where the command is run on multiple nodes concurrently, multiple errors could happen and it's useful to output error messages for all of them because they may be different. That means the messages need to be output in this method. ... Writing this up I see it making logical sense now to log the details upon exit, not log the message here, and log the message for failures when running on multiple nodes. What do you think of that approach?
___
*[pkg/cmd/roachprod/install/cluster_synced.go, line 529 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2o5Wl4218zP8L3Y4Au:-M2occzFBE86aW1pdAA6:b-r4l5ks) ([raw file](https://github.com/cockroachdb/cockroach/blob/19a53c71aed1894b05f5d974bd38aa5832bf77aa/pkg/cmd/roachprod/install/cluster_synced.go#L529)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

same comment as above
</blockquote></details>

I moved the generation of this information up about 7 lines in this method so the output gets incorporated as part of the per-node output.

Noted above the reasons for why to print these out.

There is already context about which node the error was running on (see a couple lines up: `fmt.Fprintf(stdout, "  %2d: %s\n", nodes[i], r)`).
___
*[pkg/storage/enginepb/mvcc3.pb.go, line  at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2o5_MM6y18xVsPqRTV:-M2o8-lHDVXu6vqXz4mn:buu930n) ([raw file](https://github.com/cockroachdb/cockroach/blob/19a53c71aed1894b05f5d974bd38aa5832bf77aa/pkg/storage/enginepb/mvcc3.pb.go#L)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

This seems erroneous. 
</blockquote></details>

See comment in global review comments.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
I feel better about this now. I think the linter is going to force you to add some more commentary. I'll let bob sign off. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-:-M2p3Hh5E3jp1gy3nBTj:b-ug5ye6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @bobvawter, @jlinder, and @tbg)

___
*[pkg/cmd/roachprod/errors/errors.go, line 21 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2iVrej1DJCFLg7UeMi:-M2p0XRkBep8mOI7gC6r:bo6apac) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a3090cfb1b940e78c7e2911d86b40e9a767de12/pkg/cmd/roachprod/errors/errors.go#L21)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

I think I did what you asked. Did I interpret correctly?
</blockquote></details>

Indeed.
___
*[pkg/cmd/roachprod/errors/errors.go, line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2jic9z0X6Z0noRz8sj:-M2p3A2r0qpGtuGLjgEM:b-s229uv) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a3090cfb1b940e78c7e2911d86b40e9a767de12/pkg/cmd/roachprod/errors/errors.go#L35)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

I took off the trailing `Error` from all four of the errors. Following the logic of: the package is called `errors` and the types in the package are all types of errors + they are only ever used with a reference to the errors package (or within the package itself) and have that context thus there is no need for duplicating the word 'error' in the name.
</blockquote></details>

Fair enough. 
___
*[pkg/cmd/roachprod/errors/errors.go, line 20 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2p0fPk4ENhM-hpumMP:-M2p0fPk4ENhM-hpumMQ:b-uv3g1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3570100d5937d15f879c246020829aacf7c303d/pkg/cmd/roachprod/errors/errors.go#L20)):*
> ```Go
> )
> 
> // Error: Errors implementing this interface are used by the wrap() function in
> ```

The linter is going to complain about this. The comment needs to start `Error ...`. Probably should be `Error is an interface extending the stdlib error ...`
___
*[pkg/cmd/roachprod/install/cluster_synced.go, line 508 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2o5C94DWeU7UgmG1cE:-M2p1yVkDr-NCPscalPi:b-r6297v) ([raw file](https://github.com/cockroachdb/cockroach/blob/19a53c71aed1894b05f5d974bd38aa5832bf77aa/pkg/cmd/roachprod/install/cluster_synced.go#L508)):*
<details><summary><i>Previously, jlinder (James H. Linder) wrote…</i></summary><blockquote>

Great thoughts.

Using `stderr` here is likely the right move because this is error information and not regular logging. Does roachtest capture stderr output in the correct place in the logs from roachprod runs (including placing it contextually where the message happens)? If not, we'll want to update roachtest to capture them correctly. (Because adding these messages is specifically done to be captured in the roachtest logs.) 

I added a `WithDetail` containing the node number and the command that was run. Getting that info all output somewhere seems like it would be helpful for debugging (particularly for roachtest). Since this line only gets called when the command runs on one node, outputting the error info logically would be done in the Command wrapper in roachprod main. However, for the case below where the command is run on multiple nodes concurrently, multiple errors could happen and it's useful to output error messages for all of them because they may be different. That means the messages need to be output in this method. ... Writing this up I see it making logical sense now to log the details upon exit, not log the message here, and log the message for failures when running on multiple nodes. What do you think of that approach?
</blockquote></details>

Yep, this is better. 

Looking at roachtest a bit, it seems like it generally does always capture stderr. 


<!-- Sent from Reviewable.io -->

jlinder reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46201)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @bobvawter, @jlinder, and @tbg)

___
*[pkg/cmd/roachprod/errors/errors.go, line 20 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M2p0fPk4ENhM-hpumMP:-M37gEXkAzwAlFFpD4Rl:b-96oqwo) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3570100d5937d15f879c246020829aacf7c303d/pkg/cmd/roachprod/errors/errors.go#L20)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

The linter is going to complain about this. The comment needs to start `Error ...`. Probably should be `Error is an interface extending the stdlib error ...`
</blockquote></details>

It did. And a couple other things. Fixed them.


<!-- Sent from Reviewable.io -->

jlinder commented:
@bobvawter PTAL

bobvawter reviewed (approved):
LGTM w/ nits.

> Reviewed 1 of 6 files at r4, 4 of 4 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-:-M38Rdme1futmOmnld5i:b-ott0jx)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jlinder and @tbg)

___
*[pkg/cmd/roachprod/errors/errors.go, line 1 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M38Px6R6e5lP3EWLqsV:-M38Px6R6e5lP3EWLqsW:bbz67gu) ([raw file](https://github.com/cockroachdb/cockroach/blob/9eb0447788ada4cd6ba1f822e47080ca1cd88bb2/pkg/cmd/roachprod/errors/errors.go#L1)):*
> ```Go
> // Copyright 2018 The Cockroach Authors.
> ```

For new files, use the current year.  Don't update pre-existing files.
___
*[pkg/cmd/roachprod/errors/errors.go, line 44 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M38Qd9fCtfL26n6WBQe:-M38Qd9fCtfL26n6WBQf:b-fd4cpo) ([raw file](https://github.com/cockroachdb/cockroach/blob/9eb0447788ada4cd6ba1f822e47080ca1cd88bb2/pkg/cmd/roachprod/errors/errors.go#L44)):*
> ```Go
> // errors.
> func (e Cmd) ExitCode() int {
> 	return 20
> ```

Move the definitions of these values to a `const()` block at the top of the file so that a casual reader can see at-a-glance where/why the values are what they are.  The const definitions don't necessarily have to be exported.
___
*[pkg/cmd/roachprod/install/cluster_synced.go, line 85 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M38RHjS4vhYqxW1gZND:-M38RHjS4vhYqxW1gZNE:bkpsqpv) ([raw file](https://github.com/cockroachdb/cockroach/blob/9eb0447788ada4cd6ba1f822e47080ca1cd88bb2/pkg/cmd/roachprod/install/cluster_synced.go#L85)):*
> ```Go
> }
> 
> // CmdKind is the kind of command passed to	SyncedCluster.Run().
> ```

Odd whitespace in this comment.


<!-- Sent from Reviewable.io -->

jlinder reviewed (commented):
Updated. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-:-M3BpZ_o7SegxankDRyn:b1lre7r)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bobvawter, @jlinder, and @tbg)

___
*[pkg/cmd/roachprod/errors/errors.go, line 1 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M38Px6R6e5lP3EWLqsV:-M3Bgj-j5ZXOZgbS5v5L:bttuwhx) ([raw file](https://github.com/cockroachdb/cockroach/blob/9eb0447788ada4cd6ba1f822e47080ca1cd88bb2/pkg/cmd/roachprod/errors/errors.go#L1)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

For new files, use the current year.  Don't update pre-existing files.
</blockquote></details>

Ah! Of course. Thanks for the catch.
___
*[pkg/cmd/roachprod/errors/errors.go, line 44 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M38Qd9fCtfL26n6WBQe:-M3Bgnit9MAQZIR5s_gq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9eb0447788ada4cd6ba1f822e47080ca1cd88bb2/pkg/cmd/roachprod/errors/errors.go#L44)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

Move the definitions of these values to a `const()` block at the top of the file so that a casual reader can see at-a-glance where/why the values are what they are.  The const definitions don't necessarily have to be exported.
</blockquote></details>

Done.
___
*[pkg/cmd/roachprod/install/cluster_synced.go, line 85 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-M38RHjS4vhYqxW1gZND:-M3Bhn4R1qDO6nN3BTml:b-g4siwe) ([raw file](https://github.com/cockroachdb/cockroach/blob/9eb0447788ada4cd6ba1f822e47080ca1cd88bb2/pkg/cmd/roachprod/install/cluster_synced.go#L85)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

Odd whitespace in this comment.
</blockquote></details>

Ha. It's a tab that showed as just one space in goland. Fixed.


<!-- Sent from Reviewable.io -->

bobvawter reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 2 of 2 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46201#-:-M3BrIdg4PyNoHV8F9RD:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @bobvawter, @jlinder, and @tbg)



<!-- Sent from Reviewable.io -->

jlinder commented:
bors r+

jlinder commented:
fixed the failing test

bors r+
